crates/sui-framework/sources/governance/staking_pool.move

```diff
@@ -271,25 +281,56 @@ module sui::staking_pool {
 
     // ==== inactive pool related ====
 
-    /// Deactivate a staking pool by wrapping it in an `InactiveStakingPool` and sharing this newly created object.
-    /// After this pool deactivation, the pool stops earning rewards. Only delegation withdraws can be made to the pool.
-    public(friend) fun deactivate_staking_pool(pool: StakingPool, ctx: &mut TxContext) {
-        let inactive_pool = InactiveStakingPool { id: object::new(ctx), pool};
-        transfer::share_object(inactive_pool);
+    /// Deactivate a staking pool by setting the `deactivation_epoch`. After
+    /// this pool deactivation, the pool stops earning rewards. Only delegation
+    /// withdraws can be made to the pool.
+    public(friend) fun deactivate_staking_pool(pool: &mut StakingPool, ctx: &mut TxContext) {
+        // We can't deactivate an already deactivated pool.
+        assert!(pool_active(pool), EDeactivationOfInactivePool);
+        pool.deactivation_epoch = option::some(tx_context::epoch(ctx));
+    }
+
+    public(friend) fun destroy_empty_pool(pool: StakingPool) {
+        assert!(can_be_destroyed(&pool), EDestroyNonzeroBalance);
+        let StakingPool {
+            id,
+            starting_epoch: _,
+            deactivation_epoch: _,
+            sui_balance: _,
+            pending_total_sui_withdraw: _,
+            pending_pool_token_withdraw: _,
+            rewards_pool,
+            pool_token_balance: _,
+            exchange_rates,
+            pending_delegation: _,
+        } = pool;
+        balance::destroy_zero(rewards_pool);
+        object::delete(id);
+        table::drop(exchange_rates);
+    }
+
+    /// A staking pool can be destroyed if is is inactive and there is no stake outstanding.
+    public(friend) fun can_be_destroyed(pool: &StakingPool): bool {
+        !pool_active(pool) && balance::value(&pool.rewards_pool) == 0 && pool.sui_balance == 0 &&
```

Comment:
This doesn't work because of integer divisions. It's almost guaranteed that there will be some dust sui in this pool.

Reply:
Ah good point. In that case I'll remove this, no sense in keeping the extra logic around :)

Reply:
Removing this sounds good now. Later we can add logic to gc the pool and transfer the dust SUI into storage fund when `pool.pool_token_balance` is 0.